### PR TITLE
Create new AbstractTransportTest for Transports

### PR DESCRIPTION
**Pull request by ejona86**
This is lower-level than the existing AbstractTransportTest. It should
be used by all transports, but InProcessTransport is the only one as
part of this commit.

This is the main commit of #1397. Since #1397 is so large covering so many detailed areas, I'm splitting this out. After it is in, I will create separate PRs for the okhttp and netty changes.

@nmittler and @zhangkun83, no real rush on reviewing this, but it would be good to get in as it is a good place for regression tests and more precise than interop-testing's AbstractTransportTest..

### CODE REVIEW

File: core/build.gradle [hunk 1]
```diff
@@ -7,6 +7,7 @@ description = 'gRPC: Core'
 dependencies {
     compile libraries.guava,
             libraries.jsr305
+    testCompile project(':grpc-testing')
```
Comment (nmittler) [inline core/build.gradle]: This would suck in Maven :)

File: core/src/main/java/io/grpc/inprocess/InProcessTransport.java [hunk 2]
```diff
@@ -100,6 +100,7 @@ public void run() {
       shutdownThread.setDaemon(true);
       shutdownThread.setName("grpc-inprocess-shutdown");
       shutdownThread.start();
+      return;
```
Comment (nmittler) [inline core/src/main/java/io/grpc/inprocess/InProcessTransport.java]: Why did you add this?

File: testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java [hunk 3]
```diff
@@ -0,0 +1,907 @@
...
+    server.start(serverListener);
+    InOrder inOrder = inOrder(mockClientTransportListener);
+    client.start(mockClientTransportListener);
+    client.shutdown();
+    verify(mockClientTransportListener, timeout(1000)).transportTerminated();
+    inOrder.verify(mockClientTransportListener).transportShutdown(any(Status.class));
+    inOrder.verify(mockClientTransportListener).transportTerminated();
+    server.shutdown();
+    assertTrue(serverListener.waitForShutdown(1000, TimeUnit.MILLISECONDS));
+    server = null;
```
Comment (nmittler) [inline testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java]: Should we just do this in teardown?

Comment (nmittler) [inline testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java]: oh I see, you're avoiding having teardown shut it down.  Maybe add a comment?

Comment (nmittler) [inline testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java]: Actually, why have this code here at all if we'll just do it in `teardown`?

File: testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java [hunk 4]
```diff
@@ -0,0 +1,907 @@
...
+    server.start(serverListener);
+    client.start(mockClientTransportListener);
+    MockServerTransportListener serverTransportListener
+        = serverListener.takeListenerOrFail(1000, TimeUnit.MILLISECONDS);
+    serverTransport = serverTransportListener.transport;
+
+    ClientStream clientStream = client.newStream(methodDescriptor, new Metadata());
+    clientStream.start(mockClientStreamListener);
+    StreamCreation serverStreamCreation
+        = serverTransportListener.takeStreamOrFail(1000, TimeUnit.MILLISECONDS);
```
Comment (nmittler) [inline testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java]: You use `1000` a lot ... maybe use a constant? TIMEOUT?

File: testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java [hunk 5]
```diff
@@ -0,0 +1,907 @@
...
+
+  @Test
+  public void ping() throws Exception {
+    server.start(serverListener);
+    client.start(mockClientTransportListener);
+    ClientTransport.PingCallback mockPingCallback = mock(ClientTransport.PingCallback.class);
+    try {
+      client.ping(mockPingCallback, MoreExecutors.directExecutor());
+    } catch (UnsupportedOperationException ex) {
+      assumeTrue(false);
```
Comment (nmittler) [inline testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java]: Maybe a comment here to indicate what you're doing?

File: testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java [hunk 6]
```diff
@@ -0,0 +1,907 @@
...
+    client.start(mockClientTransportListener);
+    // Stream prevents termination
+    ClientStream stream = client.newStream(methodDescriptor, new Metadata());
+    stream.start(mockClientStreamListener);
+    client.shutdown();
+    ClientTransport.PingCallback mockPingCallback = mock(ClientTransport.PingCallback.class);
+    try {
+      client.ping(mockPingCallback, MoreExecutors.directExecutor());
+    } catch (UnsupportedOperationException ex) {
+      assumeTrue(false);
```
Comment (nmittler) [inline testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java]: same here.

File: testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java [hunk 7]
```diff
@@ -0,0 +1,907 @@
...
+  @Test
+  public void ping_afterTermination() throws Exception {
+    server.start(serverListener);
+    client.start(mockClientTransportListener);
+    client.shutdown();
+    ClientTransport.PingCallback mockPingCallback = mock(ClientTransport.PingCallback.class);
+    try {
+      client.ping(mockPingCallback, MoreExecutors.directExecutor());
+    } catch (UnsupportedOperationException ex) {
+      assumeTrue(false);
```
Comment (nmittler) [inline testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java]: and here.

File: testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java [hunk 8]
```diff
@@ -0,0 +1,907 @@
...
+    client.start(mockClientTransportListener);
+    MockServerTransportListener serverTransportListener
+        = serverListener.takeListenerOrFail(1000, TimeUnit.MILLISECONDS);
+    serverTransport = serverTransportListener.transport;
+
+    Metadata clientHeaders = new Metadata();
+    clientHeaders.put(asciiKey, "client");
+    clientHeaders.put(asciiKey, "dupvalue");
+    clientHeaders.put(asciiKey, "dupvalue");
+    clientHeaders.put(binaryKey, "äbinaryclient");
```
Comment (nmittler) [inline testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java]: haha ... German is basically binary :)

File: core/src/main/java/io/grpc/inprocess/InProcessTransport.java (hunk 2 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/inprocess/InProcessTransport.java]: It's a bug fix found with the tests. Without the return transportReady could be called after transportShutdown and transportTerminated. In addition, it is wrong to call transportReady if we fail to "connect."

File: testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java [hunk 9]
```diff
@@ -0,0 +1,907 @@
...
+    serverHeaders.put(binaryKey, "äbinaryserver");
+    serverStream.writeHeaders(serverHeaders);
+    verify(mockClientStreamListener, timeout(1000)).headersRead(metadataCaptor.capture());
+    assertEquals(Lists.newArrayList(serverHeaders.getAll(asciiKey)),
+        Lists.newArrayList(metadataCaptor.getValue().getAll(asciiKey)));
+    assertEquals(Lists.newArrayList(serverHeaders.getAll(binaryKey)),
+        Lists.newArrayList(metadataCaptor.getValue().getAll(binaryKey)));
+
+    clientStream.request(1);
+    verify(mockServerStreamListener, timeout(1000)).onReady();
```
Comment (nmittler) [inline testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java]: To help reduce the verbosity, you could make a static method `verifyWithTimeout(obj)` that automatically adds the `timeout(1000)` ... WDYT?

File: testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java (hunk 3 above)
Comment (ejona86) [inline testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java]: At the time teardown didn't have the waitForShutdown. Yeah, I can remove it here.

File: testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java (hunk 4 above)
Comment (ejona86) [inline testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java]: Replaced with `TIMEOUT_MS`. I actually have little problem using constants in a test, but I had actually intended to replace those before sending it out.

File: testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java (hunk 5 above)
Comment (ejona86) [inline testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java]: Done. Ditto elsewhere.

File: core/build.gradle (hunk 1 above)
Comment (ejona86) [inline core/build.gradle]: Yeah, I see no reason to use Maven at this point :)

File: testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java [hunk 10]
```diff
@@ -0,0 +1,907 @@
...
+    ServerStream serverStream = serverStreamCreation.stream;
+    ServerStreamListener mockServerStreamListener = serverStreamCreation.listener;
+
+    Status status = Status.INVALID_ARGUMENT.withDescription("I'm not listening");
+    serverStream.close(status, new Metadata());
+    verify(mockServerStreamListener, timeout(1000)).closed(statusCaptor.capture());
+    assertCodeEquals(Status.OK, statusCaptor.getValue());
+    verify(mockClientStreamListener, timeout(1000))
+        .closed(statusCaptor.capture(), any(Metadata.class));
+    assertEquals(status.getCode(), statusCaptor.getValue().getCode());
```
Comment (nmittler) [inline testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java]: Should we verify that the description was propagated as well?

File: testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java (hunk 8 above)
Comment (ejona86) [inline testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java]: :smile: Über is my typical test string, but it didn't align with how my values were descriptive.

**Comment by ejona86** [comment]
@nmittler PTAL

File: testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java (hunk 9 above)
Comment (ejona86) [inline testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java]: I find that to be an antipattern. For me, "helper methods" in tests tend to make the tests hard to read. I've also seen a lot of broken tests because they reused a helper method that was inappropriate. There are times it is necessary because of how much boilerplate is necessary or because the method name is more descriptive than some obscure code, but in general I would strongly prefer the duplication.

File: testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java [hunk 11]
```diff
@@ -0,0 +1,907 @@
...
+    // Presently we can't sent much back to the client in this case. Verify that is the current
+    // behavior for consistency between transports.
+    assertCodeEquals(Status.CANCELLED, statusCaptor.getValue());
+
+    // Second cancellation shouldn't trigger additional callbacks
+    reset(mockServerStreamListener);
+    reset(mockClientStreamListener);
+    serverStream.cancel(status);
+    verify(mockServerStreamListener, never()).closed(any(Status.class));
+    verify(mockClientStreamListener, never()).closed(any(Status.class), any(Metadata.class));
```
Comment (nmittler) [inline testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java]: Do you need to specify a timeout?

File: testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java [hunk 12]
```diff
@@ -0,0 +1,907 @@
...
+      @Override
+      public Void answer(InvocationOnMock invocation) throws Exception {
+        Object[] args = invocation.getArguments();
+        ((InputStream) args[0]).close();
+        return null;
+      }
+    };
+
+    String largeMessage;
+    {
```
Comment (nmittler) [inline testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java]: why the extra code block?

File: testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java [hunk 13]
```diff
@@ -0,0 +1,907 @@
...
+
+    doAnswer(closeStream).when(mockServerStreamListener).messageRead(any(InputStream.class));
+    serverStream.request(1);
+    verify(mockClientStreamListener, timeout(1000)).onReady();
+    assertTrue(clientStream.isReady());
+    final int maxToSend = 10 * 1024;
+    int clientSent;
+    for (clientSent = 0; clientStream.isReady(); clientSent++) {
+      if (clientSent > maxToSend) {
+        fail("Too many messages sent before isReady() returned false");
```
Comment (nmittler) [inline testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java]: Can you clarify how this works? Maybe we need a comment?

File: testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java [hunk 14]
```diff
@@ -0,0 +1,907 @@
...
+      }
+      clientStream.writeMessage(methodDescriptor.streamRequest(largeMessage));
+      clientStream.flush();
+    }
+    assertTrue(clientSent > 0);
+    for (; clientSent < 5; clientSent++) {
```
Comment (nmittler) [inline testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java]: Again, some clarification would be helpful .... isn't `clientSent` going to be > 5?

File: testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java (hunk 11 above)
Comment (ejona86) [inline testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java]: A timeout would be the equivalent of a sleep (and I don't know if you can do a timeout with never()). I'm adding a `doPingPong()` and some Javadocs to the `doPingPong()` method.

**Comment by nmittler** [comment]
@ejona86 finished first round

File: testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java (hunk 10 above)
Comment (ejona86) [inline testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java]: Yeah. Generally I don't want to check description every place I could, just places it seems more likely for something to go wrong. But there isn't another test verify the status in this case so it does make sense to verify the description.

File: testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java (hunk 12 above)
Comment (ejona86) [inline testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java]: Readability. The local variables are thrown away and the side-effects of the code are limited. It allows you to easily ignore the block.

File: testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java (hunk 14 above)
Comment (ejona86) [inline testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java]: I'm adding a comment:
> Make sure there are at least a few messages buffered.

If `isReady()` returns `false` almost-immediately (as it will actually do for InProcessTransport), then we want at least a couple messages.

File: testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java (hunk 13 above)
Comment (ejona86) [inline testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java]: Adding a comment.

**Comment by ejona86** [comment]
@nmittler PTAL

File: core/src/main/java/io/grpc/inprocess/InProcessTransport.java (hunk 2 above)
Comment (nmittler) [inline core/src/main/java/io/grpc/inprocess/InProcessTransport.java]: Oh I see ... was reading the diff wrong (below was collapsed).

File: testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java [hunk 15]
```diff
@@ -0,0 +1,927 @@
...
+      largeMessage = sb.toString();
+    }
+
+    doAnswer(closeStream).when(mockServerStreamListener).messageRead(any(InputStream.class));
+    serverStream.request(1);
+    verify(mockClientStreamListener, timeout(TIMEOUT_MS)).onReady();
+    assertTrue(clientStream.isReady());
+    final int maxToSend = 10 * 1024;
+    int clientSent;
+    for (clientSent = 0; clientStream.isReady(); clientSent++) {
```
Comment (nmittler) [inline testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java]: Maybe just a comment before the loop to the effect of:

`Verify that outbound flow control will push back on the application`.

File: testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java [hunk 16]
```diff
@@ -0,0 +1,927 @@
...
+    clientStream.start(mockClientStreamListener);
+
+    MockServerTransportListener serverTransportListener
+        = serverListener.takeListenerOrFail(TIMEOUT_MS, TimeUnit.MILLISECONDS);
+    StreamCreation serverStreamCreation
+        = serverTransportListener.takeStreamOrFail(TIMEOUT_MS, TimeUnit.MILLISECONDS);
+    ServerStream serverStream = serverStreamCreation.stream;
+    ServerStreamListener mockServerStreamListener = serverStreamCreation.listener;
+
+    serverStream.close(Status.OK, new Metadata());
```
Comment (nmittler) [inline testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java]: So we're just ping-ponging headers? Seems fine, just not what I expected for `pingpong`.

**Comment by nmittler** [comment]
@ejona86 LGTM

File: testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java [hunk 17]
```diff
@@ -0,0 +1,927 @@
...
+  @Test
+  public void openStreamPreventsTermination() throws Exception {
+    server.start(serverListener);
+    client.start(mockClientTransportListener);
+    MockServerTransportListener serverTransportListener
+        = serverListener.takeListenerOrFail(TIMEOUT_MS, TimeUnit.MILLISECONDS);
+    serverTransport = serverTransportListener.transport;
+
+    ClientStream clientStream = client.newStream(methodDescriptor, new Metadata());
+    clientStream.start(mockClientStreamListener);
```
Comment (zhangkun83) [inline testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java]: There is a race between transport shutdown and stream start. Shall we test that an unstarted stream would also prevent termination?

File: testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java (hunk 15 above)
Comment (ejona86) [inline testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java]: I added a comment similar to that, but said "client" or "server" and tweaked it a bit.

File: testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java (hunk 16 above)
Comment (ejona86) [inline testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java]: Consider it an optimization :-). It was actually going to send a message back and forth, but then I realized I just need to ping-pong headers.

File: testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java (hunk 17 above)
Comment (ejona86) [inline testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java]: Today, okhttp and netty are only considering started streams as preventing termination and newStream is just a constructor. That is the same case for Channel and Call; only started Calls prevent shutdown of the Channel.

`InProcessTransport` is a bit of a mix that is probably buggy, it should probably swap to basing off start(); the newStream() check really relates to work done in start(). `DelayedClientTransport` used newStream() since it was convenient.

The difference between the two approaches doesn't really matter too much in my mind because they are indecipherable in ClientCallImpl since we start() the stream very soon after creating it, and we don't have any other synchronization in place in between.

The only real import difference is what happens if you create a _call_ and never start it. For streams it really doesn't seem to matter since it would only tweak the likelihood of winning a race.
